google/cloud/bigtable/examples/stock_backtesting/backtesting_main.cc

```diff
@@ -0,0 +1,258 @@
+      cbt::Filter::TimestampRange(start_milli, end_milli);
+
+  return cbt::Filter::Chain(column_filter, timestamp_filter);
+}
+
+void CalculateProfit(
+    std::unordered_map<std::string, std::map<absl::CivilDay, double>> const&
+        time_series,
+    cbt::examples::Strategy const& strategy) {
+  // Assume the base and sample fields are the same across conditions.
+  auto base_it =
+      time_series.find(strategy.conditions(0).base())->second.begin();
+  auto sample_it =
+      time_series.find(strategy.conditions(0).sample())->second.begin();
+  ++sample_it;
+
+  double shares = 0.0, wallet = 0.0;
+  while (sample_it !=
+         time_series.find(strategy.conditions(0).sample())->second.end()) {
+    double const sample_price = sample_it->second;
+    double const base_price = base_it->second;
+    double const change = (sample_price - base_price) / base_price;
+    for (auto const& condition : strategy.conditions()) {
+      if (condition.threshold() > 0.0 && change > condition.threshold()) {
+        // Price goes up; buy in.
+        shares += condition.moneyin() / sample_price;
+        wallet -= condition.moneyin();
+      } else if (condition.threshold() < 0.0 &&
+                 change < condition.threshold()) {
+        // Pirce goes down; sell out.
+        shares -= condition.moneyin() / sample_price;
+        wallet += condition.moneyin();
+      }
+    }
+
+    ++sample_it;
+    ++base_it;
+  }
+
+  // At the last day of the backtesting period, we evaluate the total value of
+  // our investment, and calculate the profit.
+  std::cout << "Shares in hand: " << shares << " @ " << base_it->second
+            << std::endl;
+  std::cout << "Money in hand: " << wallet << std::endl;
+  std::cout << "Total profit: " << shares * base_it->second + wallet
+            << std::endl;
+}
+
+}  // namespace
+
+int main(int argc, char* argv[]) {
+  if (argc != 8) {
+    std::cerr
+        << "Usage: backtesting <strategy_filepath> <ticker> "
+        << "<start_date> <end_date> <project_id> <instance_id> <table_id>";
+    return 1;
+  }
+
+  // Prerequisite check.
+  std::string const strategy_filepath = argv[1];
+  std::string ticker = argv[2];
+  absl::AsciiStrToUpper(&ticker);
+  std::string const start_date_str = argv[3];
+  std::string const end_date_str = argv[4];
+  std::string const project_id = argv[5];
+  std::string const instance_id = argv[6];
+  std::string const table_id = argv[7];
+  if (strategy_filepath.empty() || ticker.empty() || start_date_str.empty() ||
+      end_date_str.empty() || project_id.empty() || instance_id.empty() ||
+      table_id.empty()) {
+    std::cerr << "Please specify necessary parameters.";
+    return 1;
+  }
+
+  absl::CivilDay start_date, end_date, min_start_date, max_end_date;
+  if (!absl::ParseCivilTime(start_date_str, &start_date) ||
+      !absl::ParseCivilTime(end_date_str, &end_date)) {
+    std::cerr << "Can't parse the dates: " << start_date_str << " and "
+              << end_date_str;
```

Comment:
On errors, I think you need to print something  more helpful like: "Expected date in X format"